Update fully-typed-contracts.mdx (#823)
stellar/stellar-docs: docs/learn/encyclopedia/contract-development/types/fully-typed-contracts.mdx

  />
</head>

When you compile a contract created with [soroban-sdk](../../../../tools/sdks/library.mdx#soroban-rust-sdk), the Wasm file ends up with a [custom section](https://webassembly.github.io/spec/core/appendix/custom.html) containing a machine-readable description of your contract's interface types, sometimes called its [spec](https://github.com/stellar/rs-soroban-sdk/tree/main/soroban-spec) or its [API](https://github.com/stellar/soroban-docs/pull/381#issuecomment-1507283476). This is similar to [ABIs](https://www.quicknode.com/guides/ethereum-development/smart-contracts/what-is-an-abi/) in Ethereum, except that Soroban will store every single one of them on-chain from day one, and they include comments from the contract's author.

> These interface types are formatted using [XDR](../../data-format/xdr.mdx), a data format used widely throughout Stellar. It can be tricky to create or consume XDR manually, but tooling can fetch these interface types to make your life easier. [Stellar CLI](../../../../tools/developer-tools/README.mdx#cli) and [Stellar SDK](../../../../tools/sdks/library.mdx#javascript-sdk) are two such tools to do so. Let's look at each.

Reviewer: ```suggestion
These interface types are formatted using [XDR](../../data-format/xdr.mdx), a data format used widely throughout Stellar. It can be tricky to create or consume XDR manually, but tooling can fetch these interface types to make your life easier. [Stellar CLI](../../../../tools/developer-tools/cli/README.mdx#cli) and [Stellar SDK](../../../../tools/sdks/library.mdx#javascript-sdk) are two such tools to do so. Let's look at each.
```

## Stellar CLI: `stellar contract invoke`

Really, every smart contract is its own program, and deserves its own CLI.

So that's what Stellar CLI gives you.

A unique CLI for each smart contract. Constructed on-the-fly, right from the on-chain interface types. Including the author's comments. An _implicit CLI_.

For example, calling the native asset contract on the Test network:

```bash

> $ soroban contract invoke --network testnet --id CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC -- --help

Reviewer: ```suggestion
$ stellar contract invoke --network testnet --id CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC -- --help
```

:::tip

If you're unfamiliar with the `--` double dash separator, this is a pattern used by other CLIs. Everything after the double dash, sometimes called the [slop](https://github.com/clap-rs/clap/issues/971), gets passed to the child process. An example of other CLIs that make use of this are `npm run` and `cargo run`.

:::

## Stellar JS SDK: `contract.Client`

To create a contract client for the same contract as shown for `contract invoke` above, you would use this JavaScript:

```js
import { contract } from "@stellar/stellar-sdk";
const xlm = contract.Client({
  contractId: "CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC",
  networkPassphrase: "Test SDF Network ; September 2015",
  rpcUrl: "https://soroban-testnet.stellar.org",
});
```

Like the CLI, this will fetch the contract from the live network and parse the contract's types/spec. It will auto-generate a class that allows you to ergonomically call the contract's methods:

```js
xlm.balance({ id: "G123…" });
```

> Note that everything shown above works dynamically from a browser. However, sometimes it's nice to also have this behavior available as a library, while building apps. And in this case, it's really nice to have TypeScript, which enables type-ahead for all methods in the contract, and which can include the comments from the contract's original author. That's what the CLI's `contract bindings typescript` is for:

Reviewer: ```suggestion
Note that everything shown above works dynamically from a browser. However, sometimes it's nice to also have this behavior available as a library, while building apps. And in this case, it's really nice to have TypeScript, which enables type-ahead for all methods in the contract, and which can include the comments from the contract's original author. That's what the CLI's `contract bindings typescript` command is for:
```

```bash
stellar contract bindings typescript \
  --output-dir xlm \
  --id CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC \

>   --overwrite

Reviewer: Don't we need `--network testnet` somewhere in here? Or, if we're trying to illustrate _just_ the unique parts of the bindings command, I'm fine with it as-is

Author: Yes we do, good call

:::

## Stellar JS SDK: `contract.Client`

To create a contract client for the same contract as shown for `contract invoke` above, you would use this JavaScript:

```js
import { contract } from "@stellar/stellar-sdk";
const xlm = contract.Client({
  contractId: "CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC",
  networkPassphrase: "Test SDF Network ; September 2015",
  rpcUrl: "https://soroban-testnet.stellar.org",
});
```

Like the CLI, this will fetch the contract from the live network and parse the contract's types/spec. It will auto-generate a class that allows you to ergonomically call the contract's methods:

```js
xlm.balance({ id: "G123…" });
```

Note that everything shown above works dynamically from a browser. However, sometimes it's nice to also have this behavior available as a library, while building apps. And in this case, it's really nice to have TypeScript, which enables type-ahead for all methods in the contract, and which can include the comments from the contract's original author. That's what the CLI's `contract bindings typescript` command is for:

```bash

> stellar contract bindings typescript \
>   --network testnet
>   --output-dir xlm \
>   --id CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC \
>   --overwrite

Reviewer: ```suggestion
stellar contract bindings typescript \
  --network testnet \
  --output-dir xlm --overwrite \
  --id CDLZFC3SYJYDZT7K67VZ75HPJVIEUVNIXF47ZG2FB2RMQQVU2HHGCYSC
```

  -h, --help  Print help
```

> Like any other CLI, you can also get help for any of these subcommands using something like `… -- balance --help`. Stellar CLI again fetches the on-chain interface types, this time using it to generate a full list of all arguments to the function, and even generates examples.

Reviewer 2: Is it normal to have the "..." here? It might be! Just caught my eye.

Author: The full command this is v long, and I'm abbreviating it here. If that's confusing, I can explain it explicitly. Maybe:

```suggestion
Like any other CLI, you can also get help for any of these subcommands. Omitting everything before the `--` in the previous command, this would look like: `… -- balance --help`. Stellar CLI again fetches the on-chain interface types, this time using it to generate a full list of all arguments to the function, and even generates examples.
```

:::tip

If you're unfamiliar with the `--` double dash separator, this is a pattern used by other CLIs. Everything after the double dash, sometimes called the [slop](https://github.com/clap-rs/clap/issues/971), gets passed to the child process. An example of other CLIs that make use of this are `npm run` and `cargo run`.

:::

<!-- The following paragraph is also outdated, but I'm not sure the most up-to-date PRs/Issues/Discussions to link to. Maybe we can just remove this paragraph? -->

> Of course, the exact way that that Stellar CLI parses arguments is an ongoing design discussion. Representing complex custom arguments on the command line is a design challenge no blockchain CLI gets perfect yet; many, including Stellar CLI, currently use JSON. But this has downsides, and other options are possible. [How would you like to see it work?](https://docs.google.com/document/d/1LASaLyxIA2-YqLnn2NqI5NZGnEmdzx9lLh-j7n0-i4A/edit#)

Author: Re-adding my comment here, to make it easier to read:

> The following paragraph is also outdated, but I'm not sure the most up-to-date PRs/Issues/Discussions to link to. Maybe we can just remove this paragraph?

Reviewer: I would think the stellar/stellar-cli repo would be the relevant resource to link to, if we wanted. But, I think removing the paragraph is probably the best route.

Author: Done
